### ISSUE TITLE
Add support for 4-bit weights in QLinear

### ISSUE THREAD

**Issue by dacorvo** [huggingface/quanto#65]
Further reducing the bitwidth of quantized tensors would allow to save even more device memory.

Unlike weights, activations are quantized dynamically during inference using predefined quantization parameters (the predefined parameters are established during calibration for efficiency: evaluating the activations range on every sample is way too costly).

This makes activations typically much harder to quantize than weights, because no optimization algorithm can be applied dynamically.

As a matter of fact it has already been proven quite hard to quantize activations to 8-bit without losing accuracy (smoothing techniques aside). For that reason, quantizing activations to 4-bit seems impossible.

This issue is therefore about the 4-bit quantization of weights only.

This raises at least two issues with the current QTensor subclass:

- it assumes its inner data is a tensor with a valid dtype (and int4 is not),
- it only supports symmetrical quantization, which would probably makes it hard to quantize to 4-bit.

Instead of tweaking QTensor directly, I would be inclined to create a new QIn4tTensor child class:
- whose _data follows different semantics (typically pack 4-bit weights into a torch.int32 tensor),
- that has a _zeropoint member.

Then, a single dispatched function should be supported for that new subclass: torch.nn.functional.linear.
It would unpack the weights to int8 to obtain a QTensor, and call the QTensor dispatch.

Finally, the QLinear module should be modified to store its weights as QInt4Tensor instead of Tensor.

**Comment by dacorvo** [comment]
I think looking at what has been done in https://github.com/pytorch-labs/gpt-fast would provide a great starting point.

In particular, there already seems to be a packing function for int4 weights.
https://github.com/pytorch-labs/gpt-fast/blob/8c8c463f1f909890a2c0cd3f9e2ad75fff7f2b6c/quantize.py#L351

**Comment by dacorvo** [comment]
On a second thought, it is better if QInt4Tensor is a new subclass completely independent from QTensor, otherwise it will inherit the op dispatch which is completely unnecessary and error prone.

So basically it would be a streamlined version of QTensor, containing only:
- `__new__`, `__init__` and `__repr__`,
- `quantize` to create a new int4 tensor from a float tensor,
- `__torch_function__` to handle `nn.functional.linear` (raising an error on any other func),
- `__tensor_flatten` and `__tensor_unflatten__` to be compatible with dynamo.
I am not even sure we would need to expose a dequantize operation.

**Comment by dacorvo** [comment]
Preliminary implementation merged in https://github.com/huggingface/quanto/pull/68.

**Comment by dacorvo** [comment]
Next steps are:
- to store packed int4 weights (they are now stored as int8),
- to provide optimized kernels for matmul (and possibly linear).

**Comment by dacorvo** [comment]
Packed weights contributed by @younesbelkada in https://github.com/huggingface/quanto/pull/70.

Next step is to add custom kernels for packed mm.

**Comment by alejandroarmas** [comment]
Hi @dacorvo I'd love to take on the custom kernels for packed mm. I did a super rough PR #107 so that I could outline my understanding of the scope and requirements of the problem. If it's alright with you, I'd love to get feedback on what I'm understanding correctly and incorrectly before I design my solution. Thank you :D

**Comment by dacorvo** [comment]
Hi @alejandroarmas thank you very much for your interest in quanto, and congratulations for the pull-request.
You are definitely in the right direction, I will comment directly in your pull-request.
That said, @younesbelkada has just resumed the work on this yesterday:
https://github.com/huggingface/quanto/tree/try-4bit-mm
in a first step, his implementation is entirely based on pytorch ops though, so if you feel you can contribute optimized mm for specific targets, you can piggyback on his branch.

**Comment by shuokay** [comment]
Hi @dacorvo,  Can I say, QTensor only supports symmetric quantization and can be used in activation and weight quantization while QBitsTensor support unsymmetric quantization and should only be used in weight quantization by design?

**Comment by dacorvo** [comment]
> Hi @dacorvo, Can I say, QTensor only supports symmetric quantization and can be used in activation and weight quantization while QBitsTensor support unsymmetric quantization and should only be used in weight quantization by design?

Yes, you are correct.
